### PR TITLE
FIX weighted loss issue

### LINKED ISSUE #92: Weights are never passed for cross entropy in binary classification
Currently, to assign weights to the loss function, we are checking the output type, and if its binary we are making a kwargs dict with 'pos_weight': weights. This assumes that for binary classification the loss function is BCEWithLogitsLoss. However, that is not the case as by default we are using CrossEntropyLoss for binary classification and therefore the weights are never passed to this loss function.

### PR DESCRIPTION
**Pull request by ravinkohli**
Changed tests for losses and how weighted strategy is handled in the base trainer

Fixes #92

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/losses.py [hunk 1]
```diff
@@ -11,16 +11,17 @@
...
     STRING_TO_OUTPUT_TYPES, STRING_TO_TASK_TYPES, TASK_TYPES_TO_STRING
 
+
 losses = dict(classification=dict(
     CrossEntropyLoss=dict(
-        module=CrossEntropyLoss, supported_output_type=MULTICLASS),
+        module=CrossEntropyLoss, supported_output_types=(MULTICLASS, BINARY)),
     BCEWithLogitsLoss=dict(
-        module=BCEWithLogitsLoss, supported_output_type=BINARY)),
+        module=BCEWithLogitsLoss, supported_output_types=(BINARY,))),
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/training/losses.py:19]: Hey, can you add a check in so that we make sure this is a list? Or can you instead of a tuple actually make a list? It is very easy to miss the comma...  Like this can be a purest making sure everything is a list or even an in code check

File: test/test_pipeline/test_losses.py [hunk 2]
```diff
@@ -32,29 +31,24 @@ def test_get_name_error():
...
+    weights = None
+    if bool(weighted) and 'classification' in dataset_properties['task_type']:
+        strategy = get_loss_weight_strategy(loss)
+        weights = strategy(y=labels)
+        weights = torch.from_numpy(weights)
+        weights = weights.type(torch.FloatTensor)
+        kwargs = {'pos_weight': weights} if loss.__name__ == 'BCEWithLogitsLoss' else {'weight': weights}
+    loss = loss() if weights is None else loss(**kwargs)
+    score = loss(predictions, targets)
+    assert isinstance(score, torch.Tensor)
```
Comment (franchuterivera) [inline test/test_pipeline/test_losses.py:54]: Just to make sure nothing is broken, can you check the size of this score and you actually are providing a fixed target... I think the value in this case is deterministic?

**Review by franchuterivera — CHANGES_REQUESTED** [review]
Hello, everything looks great, just some question to see if it is posible to improve the testing

File: autoPyTorch/pipeline/components/training/losses.py (hunk 1 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:19]: yeah I'll make them lists, and add a check for that

File: test/test_pipeline/test_losses.py (hunk 2 above)
Comment (ravinkohli) [reply test/test_pipeline/test_losses.py:54]: in this case no it is not deterministic, I didn't add the seeding. But do you think it will be better to add the seed for all np, torch at the top of the contest file?

Comment (ravinkohli) [reply test/test_pipeline/test_losses.py:55]: I have updated the test_losses in a way that we don't need to set the seed for this test as per our discussion today

File: autoPyTorch/pipeline/components/training/losses.py [hunk 3]
```diff
@@ -11,16 +11,17 @@
 from autoPyTorch.constants import BINARY, CLASSIFICATION_TASKS, CONTINUOUS, MULTICLASS, REGRESSION_TASKS, \
     STRING_TO_OUTPUT_TYPES, STRING_TO_TASK_TYPES, TASK_TYPES_TO_STRING
 
+
 losses = dict(classification=dict(
     CrossEntropyLoss=dict(
-        module=CrossEntropyLoss, supported_output_type=MULTICLASS),
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/losses.py:16]: Can you change those names to "MultiClass" and "Binary"? Plus, it would be better to define typing variable in the file that defines MULTICLASS and BINARY. e.g. SupportedOutputTypes = Union[MultiClass, Binary]

File: autoPyTorch/pipeline/components/training/losses.py [hunk 4]
```diff
@@ -11,16 +11,17 @@
...
+        module=CrossEntropyLoss, supported_output_types=[MULTICLASS, BINARY]),
     BCEWithLogitsLoss=dict(
-        module=BCEWithLogitsLoss, supported_output_type=BINARY)),
+        module=BCEWithLogitsLoss, supported_output_types=[BINARY])),
     regression=dict(
         MSELoss=dict(
-            module=MSELoss, supported_output_type=CONTINUOUS),
+            module=MSELoss, supported_output_types=[CONTINUOUS]),
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/losses.py:22]: Can you change these dicts to NamedTuple or this [BaseDict](https://github.com/automl/Auto-PyTorch/pull/88/commits/ce1f63851c3fb1cc26db2a356d8990ada2045b28)?
By explicitly define the elements of dicts, the readability will increase and we can avoid typo.
The greatest benefit from those data structures is your working environment can suggest the keys.

File: autoPyTorch/pipeline/components/training/losses.py (hunk 3 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:16]: okay, should I change the supported output types also or you'll include it in your PR as well?

File: autoPyTorch/pipeline/components/training/losses.py (hunk 4 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:22]: I can change it later, as BaseDict is included in your PR. is it fine?

**Review by nabenabe0928 — COMMENTED** [review]
Ok, let's do that so:)

File: autoPyTorch/pipeline/components/training/losses.py (hunk 3 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/losses.py:16]: I think it would be better to do it by yourself because you are more familiar with the related codes

File: autoPyTorch/pipeline/components/training/losses.py [hunk 5]
```diff
@@ -11,16 +11,17 @@
...
         L1Loss=dict(
-            module=L1Loss, supported_output_type=CONTINUOUS)))
+            module=L1Loss, supported_output_types=[CONTINUOUS])))
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:24]: Thats true, but @franchuterivera in his older comment said he prefers them to be a list as people can miss the , that is needed to make a single element tuple i.e, (CONTINUOUS,).

File: autoPyTorch/pipeline/components/training/losses.py (hunk 3 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:16]: okay that makes sense

Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:16]: I think these supported output types is a property of the particular loss. for example, if we had our own loss functions, this information will go to the get_properties function and hence for me it does not make sense to create supported output types as a typing variable.

File: autoPyTorch/pipeline/components/training/losses.py (hunk 1 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/losses.py]: But as mentioned, if we input something like supported_output_type=[MULTICLASS], the argument is gonna be updated by this.

File: autoPyTorch/pipeline/components/training/losses.py (hunk 3 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/losses.py:16]: > these supported output types are a property of the particular loss
Hum, that is true. But at the same time, we would not know the supported types explicitly from the codes, so let's make sure we can know the supported type explicitly (doc-string or class function such as CrossEntropyLoss.is_output_type_supported() etc...)

File: autoPyTorch/pipeline/components/training/losses.py (hunk 4 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/losses.py:22]: Yes, that is fine:)

File: autoPyTorch/pipeline/components/training/losses.py (hunk 3 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:16]: yes I'll add doc string for this file.

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Could you add the names of supported losses in the doc-string of get_supported_losses as well?

File: autoPyTorch/pipeline/components/training/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/losses.py:36]: I did not understand the two approaches.

[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/losses.py:36]: Ah, maybe I misunderstood. Nothing and I will resolve the conversation.

**Review by nabenabe0928 — APPROVED** [review]
